File: 907. Sum of Subarray Minimums/sumSubarrayMins.java
class Solution {
    public int[] getNsl(int[] nums, int n) {
        int[] result = new int[n];
        Stack<Integer> s = new Stack<>();

        for (int i = 0; i < n; i++) {
            while (!s.isEmpty() && nums[s.peek()] >= nums[i]) {
                s.pop();
            }

            result[i] = s.isEmpty() ? -1 : s.peek();
            s.push(i);
        }

        return result;
    }

    public int[] getNsr(int[] nums, int n) {
        int[] result = new int[n];
        Stack<Integer> s = new Stack<>();

        for (int i = n - 1; i >= 0; i--) {
            while (!s.isEmpty() && nums[s.peek()] > nums[i]) {
                s.pop();
            }

            result[i] = s.isEmpty() ? n : s.peek();
            s.push(i);
        }

        return result;
    }

    public int sumSubarrayMins(int[] arr) {
        int n = arr.length;

        int[] nsl = getNsl(arr, n);
        int[] nsr = getNsr(arr, n);

        long mod = (long) 1e9 + 7;
        long sum = 0;
        for (int i = 0; i < n; i++) {
            long ls = i - nsl[i];
            long rs = nsr[i] - i;

            long totalWays = ls * rs;

            long totalSum = (arr[i] * totalWays) % mod;

            sum = (sum + totalSum) % mod;
        }

        return (int) sum;
    }
}

File: 907. Sum of Subarray Minimums/sumSubarrayMins.cpp
class Solution
{
public:
    vector<int> getNsl(vector<int> &nums, int n)
    {

        vector<int> result(n);
        stack<int> s;

        for (int i = 0; i < n; i++)
        {
            if (s.empty())
            {
                result[i] = -1;
            }
            else
            {
                while (!s.empty() && nums[s.top()] >= nums[i])
                    s.pop();

                result[i] = s.empty() ? -1 : s.top();
            }
            s.push(i);
        }
        return result;
    }
    vector<int> getNsr(vector<int> &nums, int n)
    {

        vector<int> result(n);
        stack<int> s;

        for (int i = n - 1; i >= 0; i--)
        {
            if (s.empty())
            {
                result[i] = n;
            }
            else
            {
                while (!s.empty() && nums[s.top()] > nums[i])
                    s.pop();

                result[i] = s.empty() ? n : s.top();
            }
            s.push(i);
        }
        return result;
    }
    int sumSubarrayMins(vector<int> &arr)
    {

        int n = arr.size();

        vector<int> nsl = getNsl(arr, n);
        vector<int> nsr = getNsr(arr, n);

        long long mod = 1e9 + 7;
        long long sum = 0;
        for (int i = 0; i < n; i++)
        {

            long long ls = i - nsl[i]; // no. of smaller element at left side
            long long rs = nsr[i] - i; // no. of smaller element at right side;

            long long totalWays = ls * rs; // totalWays no of subarrays we get

            long long totalSum = (arr[i] * totalWays);

            sum = (sum + totalSum) % mod;
        }

        return sum;
    }
};